dams-key-server/src/command/authenticate.rs

```diff
@@ -1,79 +1,192 @@
-        _client: &reqwest::Client,
-        _config: &Config,
+        request: Request<tonic::Streaming<ClientAuthenticate>>,
+        db: &Database,
+        rng: Arc<Mutex<StdRng>>,
         service: &Service,
-        _session_key: SessionKey,
-        chan: Chan<protocol::Authenticate>,
-    ) -> Result<(), anyhow::Error> {
-        let server_setup = create_or_retrieve_server_key_opaque(rng, service)
-            .context("could not find/create server key")?;
-
-        let (auth_start, chan) = chan
-            .recv()
-            .with_timeout(service.message_timeout)
-            .await
-            .context("Did not receive RegisterStart")??;
+    ) -> Result<Response<AuthenticateStream>, Status> {
+        let (tx, rx) = mpsc::channel(2);
+        let mut stream = request.into_inner();
 
-        let (credential_request, user_id) = auth_start.into_parts();
-        let server_registration = match User::find_user(&db, &user_id).await? {
-            Some(user) => user.into_server_registration(),
-            None => abort!(in chan return authenticate::Error::UserIdDoesNotExist),
-        };
-        let server_login_start_result = match ServerLogin::start(
-            rng,
-            &server_setup,
-            Some(server_registration),
-            credential_request,
-            user_id.as_bytes(),
-            ServerLoginStartParameters::default(),
-        ) {
-            Ok(server_login_start_result) => server_login_start_result,
-            Err(_) => abort!(in chan return authenticate::Error::ServerError),
+        let server_setup = {
+            let mut local_rng = rng
+                .lock()
+                .map_err(|_| Status::unavailable("Unable to access RNG"))?;
+            create_or_retrieve_server_key_opaque(&mut local_rng, service)
+                .map_err(|_| Status::aborted("could not find/create server key"))?
         };
 
-        proceed!(in chan);
+        // Clone db outside of thread to prevent lifetime errors
+        let db = db.clone();
+        // let server_setup = server_setup.clone();
 
-        let chan = chan
-            .send(server_login_start_result.message)
-            .await
-            .context("Couldn't respond with AuthStartReceived")?;
+        let _ = tokio::spawn(async move {
+            let mut server_login_result: Option<ServerLoginStartResult<OpaqueCipherSuite>> = None;
+            // Process start step
+            if let Some(message) = stream.next().await {
+                let message = message?;
+                let start_message = Self::unwrap_client_start_step(message.step)?;
+                let response =
+                    Self::handle_authenticate_start(&db, start_message, rng, &server_setup).await?;
+                server_login_result = Some(response.server_message);
+                let _ = tx
+                    .send(response.result)
+                    .await
+                    .map_err(|_| Status::aborted("Handle weird error type"));
```

Comment:
```suggestion
                    .map_err(|e| Status::aborted(e.to_string()));
```

It looks like this is just an error generated by the channel so I think it's safe to print it here.

Reply:
Resolving this one to comment on a bigger block